src/actions.ml

```diff
@@ -331,47 +329,111 @@ let bench_text ~job_info () =
-  | Ok repo_id -> (
-      Lwt_io.printl "Pushing status check for bench job."
-      <&>
-      let url = f "https://gitlab.com/coq/coq/-/jobs/%d" job_info.build_id in
-      let summary = f "## GitLab Job URL:\n[GitLab Bench Job](%s)\n" url in
-      let state = job_info.build_status in
-      let context = "bench" in
-      ( match state with
-      | "success" ->
-          let* text = bench_text ~job_info () in
-          Lwt.return_ok
-            (COMPLETED, Some SUCCESS, "Bench completed successfully", text)
-      | "failed" ->
-          let* text = bench_text ~job_info () in
-          Lwt.return_ok
-            (COMPLETED, Some NEUTRAL, "Bench completed with failures", text)
-      | "running" ->
-          Lwt.return_ok (IN_PROGRESS, None, "Bench in progress", "")
-      | "cancelled" | "canceled" ->
-          Lwt.return_ok
-          (COMPLETED, Some CANCELLED, "Bench has been cancelled", "")
-      | "created" ->
-          Lwt.return_error ("Bench job has been created, ignoring info update.")
-      | _ ->
-          Lwt.return_error (f "Unknown state for bench job: %s" state) )
+  match number with
+  | None ->
+      Lwt_io.printlf "No PR number provided for bench summary so aborting."
+  | Some number -> (
+      GitHub_queries.get_repository_id ~bot_info ~owner:gh_owner ~repo:gh_repo
       >>= function
-      | Ok (status, conclusion, description, text) ->
-          GitHub_mutations.create_check_run ~bot_info ~name:context ~status
-            ~repo_id ~head_sha:job_info.common_info.head_commit ?conclusion
-            ~title:description ~details_url:url ~summary ~text ~external_id ()
+      | Ok repo_id -> (
+          Lwt_io.printl "Pushing status check for bench job."
+          <&>
+          let url =
+            f "https://gitlab.com/coq/coq/-/jobs/%d" job_info.build_id
+          in
+          let summary = f "## GitLab Job URL:\n[GitLab Bench Job](%s)\n" url in
+          let state = job_info.build_status in
+          let context = "bench" in
+          ( match state with
+          | "success" ->
+              let* results = fetch_bench_results ~job_info () in
+              let* summary_text = bench_text results in
+              let* () =
+                bench_comment ~bot_info ~owner:gh_owner ~repo:gh_repo ~number
+                  results
+              in
+              Lwt.return_ok
+                ( COMPLETED
+                , Some SUCCESS
+                , "Bench completed successfully"
+                , summary_text )
+          | "failed" ->
+              let* results = fetch_bench_results ~job_info () in
+              let* summary_text = bench_text results in
+              let* () =
+                bench_comment ~bot_info ~owner:gh_owner ~repo:gh_repo ~number
+                  results
+              in
+              Lwt.return_ok
+                ( COMPLETED
+                , Some NEUTRAL
+                , "Bench completed with failures"
+                , summary_text )
```

Comment:
I'm really not fond of this logic where part of the action is done inside this match and part of it is done outside, using the return value. My initial suggestion was to do the action inside the match, and it still holds.

In fact, I have a further suggestion which is to put a link from the comment to the check run summary. Following it would require reversing the order in which the actions are done: the check run would need to be created first, and amended to return its URL. Then the comment could be created and use this piece of data.

Reply:
Let me try to factor things out from the match then. I agree that the flow isn't great at the moment.

Reply:
Do as you want (and maybe it will look great) but note that my suggestion was to put everything inside the match instead.

Reply:
I've put everything in the match now.